### ISSUE TITLE
Ensure everyone is reading same library files

### ISSUE THREAD

**Issue by peter-ahe-google** [dart-lang/sdk#28836]
We currently have the following ways to define the platform libraries:

* `sdk/lib/_internal/sdk_library_metadata/lib/libraries.dart`
* `sdk/lib/*.platform`
* `runtime/**/*_sources.gypi`

I'd like to combine all these in a single format, see [CL 2698003003](https://codereview.chromium.org/2698003003).

This CL is here to track the task of removing all the formats we decide to not use anymore.

**Comment by lrhn** [comment]
FVIW, I would like to see a summary of what information is used and by whom - *before* we start creating yet another format that will end up handling 95% of what we need, again.
(In other words: What problems are we actually trying to solve here?)

**Comment by peter-ahe-google** [comment]
@lrhn that's a good point. I'll include gathering a list of dependencies (who needs what) as part of this task.

**Comment by zanderso** [comment]
If fasta doesn't like YAML, and Python doesn't like YAML, I think we should have only JSON.

If the only reason is that YAML is more human-readable, I don't think that's sufficient to justify the complexity of supporting two different files with the same list. (We can put comments in a separate field as suggested [here](http://stackoverflow.com/questions/244777/can-comments-be-used-in-json), and then ignore that field when interpreting it, or implement some other suggestion from that thread. We have to do something like that anyway to support one of the requirements below and to add the copyright notice).

If I'm missing something here, and there is a better reason, then we'll need the following:
- The generated JSON files should be marked read-only.
- They should have a comment at the top saying not to edit them by hand.
- There should be a presubmit check that the .yaml file and .json file have the same contents. The warning message should explain how to fix it, ideally by executing a single python script with no command line arguments.
- To make it possible to bring up self-hosting on new platforms, none of this can assume the existence of a pre-built, or checked-in Dart SDK.

**Comment by zanderso** [comment]
Agreeing with @lrhn, I think this discussion would benefit from drawing up a design doc with the requirements, etc., before we continue landing CL's around this.

**Comment by peter-ahe-google** [comment]
@zanderso I agree that we shouldn't continue landing CLs around this issue, but I felt I had to land something to remove the hard-coded paths from inside Fasta.

I will take on this issue and write a design document, I think I know who to involve to ensure the document is comprehensive.

**Comment by sigmundch** [comment]
I just ran into this as well.

@peter-ahe-google - if the doc already exists, could you share it with me as well?

I was about to delete the *.platform files (only used by dart2js) and sdk/lib/_internal/sdk_library_metadata/lib/libraries.dart (only used by analyzer and the patch_sdk tool). I was also approaching this like Peter: replace both with a YAML that gets converted to .json, and load the json directly with these tools.

**Comment by peter-ahe-google** [comment]
No design document yet, sorry.

**Comment by sigmundch** [comment]
I can start one if you like, let me know.

**Comment by sigmundch** [comment]
Just to update on status: we are getting close to integrate on a single format:
* At this time that the CFE uses libraries.json, and indirectly so does the VM, dart2js, ddk, and flutter.
* There is a deserialization abstraction here:
https://github.com/dart-lang/sdk/blob/ccc83cf5b30f7f9ab77d3fc603b923505bc22557/pkg/front_end/lib/src/base/libraries_specification.dart#L5
* The last action item here is to migrate analyzer from using libraries.dart, to use libraries.json instead.
